code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -307,105 +262,95 @@ createNetCliEnv globalEnv@ZTGlobalEnv{..} ztCliSettings peers = liftIO $ do
-            "changePeers: preconnecting (sending req) to " <> show toConnect
-
-    -- Release sockes that were not used.
-    let dataLeft = drop (length toConnect) $ sockets `zip` adapters
-    forM_ dataLeft $ \(sock,adapter) -> Z.close sock >> adapterRelease adapter
-
--- Continues connection request (moves peer from "connecting" to
--- "connected" state).
-contConnectionReq ::
-       ZTGlobalEnv -> ZTNetCliEnv -> ZTNodeId -> ZTInternalId -> IO ()
-contConnectionReq ZTGlobalEnv{..} ZTNetCliEnv{..} nodeId iId = do
-    curTime <- getCurrentTimeMs
-    resourcesM <- atomically $ do
-        clash <- HMap.member iId <$> peersRevMap ztPeers
-        if clash
-        then pure Nothing
-        else do
-            (_,sock,adapter) <-
-                fromMaybe (error $ "contConnectionReq: cant find by nodeId") .
-                HMap.lookup nodeId <$>
-                readTVar (ztPeers ^. pivConnecting)
-            modifyTVar (ztPeers ^. pivConnected) $ HMap.insert nodeId iId
-            modifyTVar (ztPeers ^. pivConnecting) $ HMap.delete nodeId
-
-            let initHbs = HeartbeatState
-                    (zsHBIntervalMin ztCliSettings)
-                    (zsHBLivenessMax ztCliSettings)
-                    curTime
-                    False
-            modifyTVar ztHeartbeatInfo $ at nodeId .~ Just initHbs
-
-            pure $ Just (sock,adapter)
-
-    let onNothing =
-            ztLog ztCliLogging Warning $
-                "contConnectionReq: internal identity " <>
-                "collision, didn't connect peer with iid" <> show nodeId
-    let onJust (sock,adapter) = do
-            Z.close sock
-            adapterRelease adapter
-            Z.connect ztCliBack $ ztNodeIdRouter nodeId
-            Z.connect ztCliSub $ ztNodeIdPub nodeId
-            ztLog ztCliLogging Debug $
-                "contConnectionReq: successfully connected to " <>
-                show nodeId <> ", " <> show iId
-    maybe onNothing onJust resourcesM
+            "changePeers: connecting to " <> show toConnect
 
+    -- Release resources that were not used.
+    let resourcesLeft = drop (length toConnect) resources
+    forM_ resourcesLeft releasePeerRes
+  where
+    -- Reformulates peers update request in terms of current peer info
+    -- var, returning the set of add/del nodes. Doesn't change PIV itself.
+    applyUpdatePeers ::
+           PeersInfoVar
+        -> ZTUpdatePeersReq
+        -> STM ([ZTNodeId], [ZTNodeId])
+    applyUpdatePeers piv UpdatePeersReq{..} = do
+        -- Normalising the request
+        let both = L.nub $ _uprDel `L.intersect` _uprAdd
+        let add' = L.nub _uprAdd L.\\ both
+        let del' = L.nub _uprDel L.\\ both
+
+        currentPeers <- HSet.toList . hmKeys <$> readTVar piv
+
+        -- We're adding all peers asked which were not connected.
+        -- And deleting all who are not connected.
+        let toAdd = add' L.\\ currentPeers
+        let toDel = del' L.\\ currentPeers
```

Comment:
lol, it seems to be an error.
if `currentPeers == S.fromList [ZTNodeId "misha.allo.ru" 100 200]` and you ask to disconnect it, after this line `toDel` will be empty, though, you had to disconnect this peer.

it probably should be `L.intersect`

Reply:
You're right. TBH I have no idea how "disconnecting peers" may be useful in DSCP scenario, so I never actually tested it.